---
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
    toc_depth: 3
---




<!-- Javascript for zooming on figures (adapted from: https://stackoverflow.com/questions/40401680) -->

<!-- Jquery import conflicts with DT::datatable so needs to be commented here -->
<!-- <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script> -->

<style>
.zoomDiv {
  display: none;
  position: fixed;
  top: 50%;
  left: 50%;
  z-index: 50;
  transform: translate(-50%, -50%);
  background-color: #FFFFFF;
  box-shadow: 0px 0px 50px #888888;
  width: fit-content;
  max-width: 90%;
  max-height: 90%;
  overflow: auto;
}

.zoomImg {
  width: 100%;
}
</style>

<script type="text/javascript">
  $(document).ready(function() {
    $('body').prepend("<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>");
    // onClick for all img except the zoomed one and link ones (filter)
    // use 'img.zoom' and out.extra='class=\"zoom\"' in chunk to specify manually which chunk images can be zoomed
    $('img:not(.zoomImg)').filter(':not(a *)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src'));
      $('.zoomDiv').show();
    });
    // onClick function for hiding div
    $('img.zoomImg').click(function() {
      $('.zoomDiv').hide();
    });
  });
</script>




```{r rmd_setKnitrOptions, include=FALSE}
# Set hook for PNG optimization ('optipng' must be installed on system)
#knitr::knit_hooks$set(optipng = knitr::hook_optipng)

# Set chunks default options
knitr::opts_chunk$set(results='asis',
                      echo = FALSE,           # Do not show chunks code
                      message = FALSE,        # Hide R regular text output in reports
                      warning = FALSE,        # Show warning messages
                      error = TRUE,           # Do not stop generating report on errors
                      cache = FALSE,          # Disable chunk result caching
                      cache.comments = FALSE, # Changing comments does not invalidate cache
                      # fig.dim = c(5, 5),      # Size of device for R figures
                      autodep = TRUE,         # Automatic chunk dependencies if cache enabled
                      optipng = '-o2');       # PNG optimization level (2=default, 5=slow, 7=very slow)
```




```{r rmd_loadParamsAndDependencies, warning=FALSE, cache = FALSE}
### Load libraries, global functions/variables, and chunks
source( file.path( WORKING_DIR, "00_generalDeps.R"));

# Import chunks from source files
read_chunk( path=file.path( WORKING_DIR, "01_prepareData.R"));
read_chunk( path=file.path( WORKING_DIR, "02_compareMonitoredGenesExpression.R"));
read_chunk( path=file.path( WORKING_DIR, "03_compareClusterExpression.R"));
read_chunk( path=file.path( WORKING_DIR, "04_examineClusterExpression.R"));
read_chunk( path=file.path( WORKING_DIR, "05_functionalEnrichment_KEGG.R"));
read_chunk( path=file.path( WORKING_DIR, "06_pathwayAnalysis_KEGG.R"));
read_chunk( path=file.path( WORKING_DIR, "90_finalize.R"));
```


---
title: `r paste( GLOBAL_DESCRIPTION, ANALYSIS_STEP_NAME, sep ="<BR>")`
---




```{r rmd_initFunctions}

### Create output directory
dir.create( PATH_ANALYSIS_OUTPUT, showWarnings = FALSE);


### Record starting time (for report and file names)
startTime = Sys.time();
startTimeReport   = format( startTime, "%Y %m %d %H:%M:%S")
startTimeFileName = format( startTime, "%Y_%m_%d_%H_%M_%S")


### Initialize utils

# Create a hash for all parameters
paramsHash = digest( environment(), algo="murmur32");

# Create a 'safe' unique prefix for output files (now specified at project scale in globalParams.R)
#outputFilesPrefix = path_sanitize( gsub("([[:punct:]])|\\s+",
#                                        "_",
#                                        paste0(PROJECT_NAME, "_",
#                                               SAMPLE_NAME, "_",
#                                               startTimeFileName, "_",
#                                               paramsHash, "_")),
#                                   replacement = "_");

# Set seed for pseudo random number generator
set.seed(SEED);

# Set the use of multiple cores using 'future' library, as implemented in Seurat3 when possible
# https://github.com/satijalab/seurat/issues/457
plan(strategy = "multicore", workers = NBCORES);

```




# Compare datasets




## Run informations {.tabset .tabset-fade}

### Run
<center>
```{r rmd_runInfo_run}
pander( c( 'Date'     = startTimeReport,
           'Sample'   = EXPERIMENT_NAME,
           'Hash'     = paramsHash));
```
</center>

### Parameters
```{r rmd_runInfo_params}
listParams = as.list( paramsEnv);
pander(listParams[order(names(listParams))]); # Show params in alphabetical order
```

### System
<center>
```{r rmd_runInfo_sysInfo}
showSimpleDT( Sys.info())
```
</center>

### R
```{r rmd_runInfo_r}
pander( sessionInfo());
```

<!-- Ending the tabset section -->
## {.toc-ignore}
<br /><br />


## Load data

```{r load_data}
<<loadData>>
```

## Compare monitored genes between conditions
```{r rmd_heterogeneity_monitoredGenes, warning = TRUE}
<<heterogeneity_monitoredGenes>>
```

### Monitored genes expression {.tabset .tabset-fade}

#### UMAP expression plot {.tabset .tabset-fade}
```{r rmd_heterogeneity_monitoredGenes_expression_projection_UMAP, fig.dim = c(4, 4), out.width='25%'}
useReduction='umap'

cat("\n<a href='#clusters-projection'>");
<<heterogeneity_dimReduc_thumbnail>>
cat("</a> \n \n");

<<heterogeneity_monitoredGenes_expression_projection>>

rm("useReduction");
```

#### UMAP expression density plot {.tabset .tabset-fade}
```{r rmd_heterogeneity_monitoredGenes_expression_density_UMAP, fig.dim = c(4, 4), out.width='25%'}
useReduction='umap'

cat("\n<a href='#clusters-projection'>");
<<heterogeneity_dimReduc_thumbnail>>
cat("</a> \n \n");

<<heterogeneity_monitoredGenes_density_projection>>

rm("useReduction");
```

#### Expression by clusters violin plot {.tabset .tabset-fade}
```{r rmd_heterogeneity_monitoredGenes_expression_violin, fig.dim = c(4, 4), out.width='25%'}
cat("\n<a href='#clusters-projection'>");
useReduction='umap'; #c( 'umap', 'tsne');
<<heterogeneity_dimReduc_thumbnail>>
rm("useReduction");
cat("</a> \n \n");

<<heterogeneity_monitoredGenes_expression_violin>>
```



## Compare gene expression in cluster between WT and KO samples (Wicoxon method) {.tabset .tabset-fade}

In this section we take the clusters of cells computed on all the cells (WT + KO) and compute for each cluster the genes that are differentially expressed in the cluster between samples. The DEG are computed using the Wilcoxon test.

The results show the3 heatmaps:
* the heatmap of the mean of expression of the DEG in the clusters fo WT cells
* the heatmap of the mean of expression of the DEG in the clusters fo KO cells
* The heatmap of the difference of the mean of expression of the DEG in the clusters (KO - WT) 


```{r compareClusterExpressionWT}
DEG_TEST_USE = "wilcox"
<<compare_cluster_expression>>
```


## Compare gene expression in cluster between WT and KO samples (DESseq2 method) {.tabset .tabset-fade}

```{r compareClusterExpressionDESeq2}
DEG_TEST_USE = "DESeq2"
<<compare_cluster_expression>>
```

## Examine specific genes expression in clusters

```{r examineClusterExpression}
<<examine_cluster_expression>>
```

## Study KEGG functional enrichment of DEG between conditions

```{r keggFunctionalEnrichment}
<<kegg_functional_enrichment>>
```

## Study the monitored KEGG pathways {.tabset .tabset-fade}

```{r keggPathwayAnalysis}
<<kegg_pathway_analysis>>
```

